spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsSendingMessageHandler.java

```diff
@@ -105,22 +152,42 @@ protected void onInit() {
 				this.destinationExpressionProcessor.setConversionService(conversionService);
 			}
 		}
+		this.evaluationContext = ExpressionUtils.createStandardEvaluationContext(getBeanFactory());
 	}
 
 	@Override
-	protected void handleMessageInternal(final Message<?> message) throws Exception {
-		if (message == null) {
-			throw new IllegalArgumentException("message must not be null");
-		}
+	protected void handleMessageInternal(final Message<?> message) {
 		Object destination = this.determineDestination(message);
 		Object objectToSend = (this.extractPayload) ? message.getPayload() : message;
 		MessagePostProcessor messagePostProcessor = new HeaderMappingMessagePostProcessor(message, this.headerMapper);
+
+		if (this.jmsTemplate instanceof DynamicJmsTemplate && this.jmsTemplate.isExplicitQosEnabled()) {
+			Integer priority = StaticMessageHeaderAccessor.getPriority(message);
+			if (priority != null) {
+				DynamicJmsTemplateProperties.setPriority(priority);
+			}
+			if (this.deliveryModeExpression != null) {
+				Integer deliveryMode =
+						this.deliveryModeExpression.getValue(this.evaluationContext, message, Integer.class);
+
+				if (deliveryMode != null) {
+					DynamicJmsTemplateProperties.setDeliveryMode(deliveryMode);
+				}
+			}
+			if (this.timeToLiveExpression != null) {
+				Long timeToLive = this.timeToLiveExpression.getValue(this.evaluationContext, message, Long.class);
+				if (timeToLive != null) {
+					DynamicJmsTemplateProperties.setTimeToLive(timeToLive);
+				}
+			}
+		}
 		try {
-			DynamicJmsTemplateProperties.setPriority(new IntegrationMessageHeaderAccessor(message).getPriority());
-			this.send(destination, objectToSend, messagePostProcessor);
+			send(destination, objectToSend, messagePostProcessor);
 		}
 		finally {
 			DynamicJmsTemplateProperties.clearPriority();
+			DynamicJmsTemplateProperties.clearDeliveryMode();
+			DynamicJmsTemplateProperties.clearTimeToLive();
 		}
```

Comment:
I wonder if it would be better to have one ThreadLocal with a container?

Reply:
Not sure what is your concern: we have a container there and it is a `ThreadLocalMap`:
```
    public void set(T value) {
        Thread t = Thread.currentThread();
        ThreadLocalMap map = getMap(t);
        if (map != null)
            map.set(this, value);
        else
            createMap(t, value);
    }
```

So, essentially with your suggestion we are going to have only a single entry in this map, but we would create that extra container anyway.
And at the same time we would lose a flexibility when we deal only with one property. For example a `setPriority()` is used from the `AbstractJmsChannel`, too.

So, what am I missing staying with the standard Java approach?

Thanks

Reply:
Well, we now have 4 `ThreadLocal`s, 4 `ThreadLocalMap`s.

I was just suggesting one `ThreadLocal<SomeContainer>` and one `DynamicJmsTemplateProperties.clear()` method.

On the other hand, we'd create/remove a new Object for each send, so I suppose the current is better (since the `ThreadLocalMap` is retained).

So, OK, LGTM.

Reply:
Huh? No, there is a single map-per-thread: `ThreadLocalMap map = getMap(t);`. As I said: essentially it is the same container as you suggest.

Reply:
I didn't mean to say there wasn't a map per thread. My main point was we could replace 

```java
DynamicJmsTemplateProperties.clearPriority();
DynamicJmsTemplateProperties.clearDeliveryMode();
DynamicJmsTemplateProperties.clearTimeToLive();
```

with just one `DynamicJmsTemplateProperties.clear()` method.

But, as I said, I am ok with this.